### ISSUE TITLE
Cannot authenticate to Github repo during the installation process

### ISSUE THREAD

**Issue by ematipico** [pnpm/pnpm#2933]
When trying to install my packages using a private registry, `pnpm` goes in error.
I think I know what's going on.

Doing the same via `npm` works fine because it installs the dependencies that are not in scope from the npm registry and installs the scoped dependencies (the private ones) from the github registry. I could verifying it by checking its `package-lock.json`

On the other hand, `pnpm` tries to install ALL the dependencies from one single registry, which is github.

### Code to reproduce the issue:

Code inside `.npmrc`

```
registry=https://npm.pkg.github.com/{SCOPE}
```

I can't give the information of the log because it contains sensitive information.

### Expected behavior:

It should install all the packages.

### Actual behavior:

```
 ERROR  GET https://npm.pkg.github.com/{SCOPE}/bootstrap: Unauthorized - 401
No authorization header was set for the request.

These authorization settings were found:
//npm.pkg.github.com/:_authToken={TOKEN}
```

### Additional information:

Changing `.npmrc` to the following works fine

```
{SCOPE}:registry=https://npm.pkg.github.com/{SCOPE}
```

**Comment by juanpicado** [comment]
You might need to include the `always-auth=true` in your `.npmrc` file.

**Comment by juanpicado** [comment]
I will triage this one.

**Comment by zkochan** [comment]
But this seems like the correct behavior. When you set the registry setting, it sets the registry for ALL packages. Maybe npm is falling back to the default registry if the one specified in the registry setting is failing?

**Comment by ematipico** [comment]
I agree with you that setting the registry for ALL the packages should lead to this behaviour (which is correct). 

Although, as you said, npm falls back to the default registry which is not correct IMHO because it could lead to security issue. I am happy to close the issue but I think it's worth having a small paragraph inside the documentation where this behaviour is explained. 

Yarn behaves in the same way.

**Comment by zkochan** [comment]
Maybe we could also add more information to the error message.

**Comment by ematipico** [comment]
Yes, that would help too I think! Maybe a suggestion to add the package scope in the `registry` key inside the `.npmrc`

**Comment by zkochan** [comment]
Are the GitHub registry docs suggesting to use the registry setting?

**Comment by juanpicado** [comment]
Yea it does 

https://docs.github.com/en/free-pro-team@latest/packages/using-github-packages-with-your-projects-ecosystem/configuring-npm-for-use-with-github-packages

**Comment by zkochan** [comment]
That is bad. I'll suggest an edit.

**Comment by juanpicado** [comment]
@ematipico could you try with the `npm@7.x`?  I guess you tried only with the latest.

**Comment by ematipico** [comment]
@juanpicado I have tried with npm 6 (can't recall witch version). I will try with npm 7 once I get the chance and let you know how it behaves.

**Comment by statianzo** [comment]
Tested this out, it looks like `https://npm.pkg.github.com/{OWNER}/lodash` is deferring to registry.npmjs.org for unknown packages. It's not the npm cli falling back.

Notice the 200 when fetching lodash from `npm.pkg.github.com` and then the followup to `registry.npmjs.org`

```
npm info it worked if it ends with ok
npm verb cli [
npm verb cli   '/usr/local/bin/node',
npm verb cli   '/usr/local/bin/npm',
npm verb cli   'i',
npm verb cli   '--no-color',
npm verb cli   '-ddd',
npm verb cli   'lodash'
npm verb cli ]
npm info using npm@6.14.8
npm info using node@v12.18.4
npm verb npm-session e8be7194d641a4ef
npm sill install loadCurrentTree
npm sill install readLocalPackageData
npm http fetch GET 200 https://npm.pkg.github.com/statianzo/lodash 489ms
npm http fetch GET 200 https://registry.npmjs.org/lodash/-/lodash-4.17.20.tgz 150ms
npm sill pacote tag manifest for lodash@latest fetched in 655ms
npm timing stage:loadCurrentTree Completed in 674ms
npm sill install loadIdealTree
npm sill install cloneCurrentTreeToIdealTree
npm timing stage:loadIdealTree:cloneCurrentTree Completed in 1ms
npm sill install loadShrinkwrap
npm timing stage:loadIdealTree:loadShrinkwrap Completed in 0ms
npm sill install loadAllDepsIntoIdealTree
npm sill resolveWithNewModule lodash@4.17.20 checking installable status
npm timing stage:loadIdealTree:loadAllDepsIntoIdealTree Completed in 5ms
npm timing stage:loadIdealTree Completed in 7ms
npm sill currentTree regtest@1.0.0
npm sill idealTree regtest@1.0.0
npm sill idealTree └── lodash@4.17.20

...
```

Resulting json when making an authenticated request to `https://npm.pkg.github.com/{OWNER}/lodash`  notice the `tarball` points to `registry.npmjs.org`

```json5
{
  "_id": "lodash",
  "_rev": "2544-0b122902f747bccd1e1b616357c0f55b",
  "name": "lodash",
  "description": "Lodash modular utilities.",
  "dist-tags": {
    "latest": "4.17.20"
  },
  "versions": {
    // ... snip
    "4.17.20": {
      // ... snip
      "dist": {
        "integrity": "sha512-PlhdFcillOINfeV7Ni6oF1TAEayyZBoZ8bcshTHqOYJYlrqzRK5hagpagky5o4HfCzzd1TRkXPMFq6cKk9rGmA==",
        "shasum": "b44a9b6297bcb698f1c51a3545a2b3b368d59c52",
        "tarball": "https://registry.npmjs.org/lodash/-/lodash-4.17.20.tgz",
        "fileCount": 1049,
        "unpackedSize": 1406354,
        "npm-signature": "-----BEGIN PGP SIGNATURE-----\r\nVersion: OpenPGP.js v3.0.4\r\nComment: https://openpgpjs.org\r\n\r\nwsFcBAEBCAAQBQJfNXAiCRA9TVsSAnZWagAAtmYP/2G2ijVcDDyRacsKXn8Q\niX5zNGG+Od+xSuOXrMRG32hjB1giuXR2t8mlsJLQpFyQVgAexcr22J0oq0Kb\nUouNSYrjm6qfK/u5ZUg8lR/Q3L+QiaxsfNnS7FWCO4xqUB0FlI5rKtnq4zhH\nWpPscWw1S/0vV1tz9OvibtfiMDWw5m9AYHk7ckISHmiEMBWURWyDjStxVjmn\nIupfRuCjJdNoxdFyRMaXktbFaCqdMoaT00x5ImxTbIR2ZQdTQT7fA8l3FvvZ\nX/avXT8sqsY0gMCJqZGZITWI/6jIvoMLeE8IqPvAweX+rjHpTFbj9u+SjhNI\noLRY2Ya3bCCUM/T7ZeShMDOeNCyqaU4p3s5VWBQ7PG8FRUjtRdJXpmBf44B4\n6ew+lh2qK+P7FcIcJ4NDpo0/pek4keOpKmUyOYEsrXbnFKdXyzMxAztydu2U\nYR20ePPVsAh2dxwnTVW+jxoutB09gmM3YhgtuOEf16dSEsu47Tntd039li0A\nOSwvLK2jEiyyaeGh7nZSaNPr2Sgj5uq1DrOjR/eKOhY0nFUzdcDAjm8ocVSe\nncavD4t+VFnABdnog5Ub11luRNoOTBvIX+c/9DshlfRzGKIu6pq8SirU6P2j\n3y40bR5yzjhxVM/E2RsqAsPgSPl/1BLyO4+4MKIxhVTh259tWf3YxbWbuXo8\n5EAn\r\n=w2Gq\r\n-----END PGP SIGNATURE-----\r\n"
      },
      // ... snip
    }
  }
  // ... snip
}
```

Difference seems that pnpm isn't sending auth to the registry.

**Comment by statianzo** [comment]
Suspected reason for not sending auth: the resulting "nerf dart" made by pnpm is different than npm's because of a trailing slash.

`pnpm` uses the `nerf-dart` package. When the `uri` of the registry is passed it's `https://npm.pkg.github.com/statianzo/`

```js
function toNerfDart (uri) {
  var parsed = url.parse(uri)
  delete parsed.protocol
  delete parsed.auth
  delete parsed.query
  delete parsed.search
  delete parsed.hash

  return url.resolve(url.format(parsed), '.')
}
```

---

`npm` (6) uses its own implementation within `npm-registry-fetch`

https://github.com/npm/npm-registry-fetch/blob/d8df0b193b34dad4627db52259322f0dbf9257d2/auth.js#L49-L57

The passed url is `https://npm.pkg.github.com/statianzo`

Notice no trailing slash. It matters when getting passed to `url.resolve`

```js
url.resolve('https://npm.pkg.github.com/statianzo/', '.')
// 'https://npm.pkg.github.com/statianzo/'
url.resolve('https://npm.pkg.github.com/statianzo', '.')
// 'https://npm.pkg.github.com/'
```

**Comment by zkochan** [comment]
@statianzo but `@pnpm/config` is normalizing the registry URLs and it is adding a trailing slash to them: https://github.com/pnpm/pnpm/blob/main/packages/config/src/getScopeRegistries.ts

**Comment by cybairfly** [comment]
Any progress on this please? Haven't been able to use `pnpm` because of this for a long time. My situation:
`.npmrc`
```
@apify-packages:registry=https://npm.pkg.github.com/apify-packages
//npm.pkg.github.com/:_authToken=<token>
```
`pnpm i`
```
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/apify-packages/@apify-packages%2Ffingerprint-injector: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
@apify-packages:registry=https://npm.pkg.github.com/apify-packages
//npm.pkg.github.com/:_authToken=72d4[hidden]
Progress: resolved 1, reused 0, downloaded 0, added 0
```
Any advice would be highly appreciated.

**Comment by marvin1up** [comment]
I have the identical issue, this does seem like a long time.
~/.npmrc
```
//npm.pkg.github.com/:_authToken=ghp_XXXXX
//npmjs.org=true
@1uphealth:registry=https://npm.pkg.github.com/1uphealth
```

then...
```
pnpm add @1uphealth/core-1upenv-script
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/1uphealth/@1uphealth%2Fcore-1upenv-script: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
//npm.pkg.github.com/:_authToken=ghp_[hidden]
@1uphealth:registry=https://npm.pkg.github.com/1uphealth
```

**Comment by damiensedgwick** [comment]
Getting a similar issue as above, pnpm won't install private package from GitHub repo despite fiddling with configs and .npmrc files

**Comment by stefanmo** [comment]
Any news?

**Comment by zkochan** [comment]
I am not sure why you all try to include the owner in the registry URL. [The docs](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry) suggest to use:

```
@OWNER:registry=https://npm.pkg.github.com
```

Though this seems to work with npm, so I guess it should works with pnpm as well.

**Comment by damiensedgwick** [comment]
> I am not sure why you all try to include the owner in the registry URL. The docs suggest to use:

@zkochan conflicting sources most likely. I can confirm your solution does not work when using pnpm on Amplify. It tries to install all packages from one repository and ignores any private ones that are in use.

**Comment by zkochan** [comment]
Try pnpm v6.23.2

**Comment by cjanietz** [comment]
The issue is still present, manually appending the / to the registry fixes it.
It keeps losing the last path segment otherwise.

**Comment by damiensedgwick** [comment]
> Try pnpm v6.23.2

will give this a try and come back to you.

**Comment by vivek-medly** [comment]
> The issue is still present, manually appending the / to the registry fixes it. It keeps losing the last path segment otherwise @zkochan

@cjanietz can you share how appending `/` fixed it? I'm running v6.23.2 and tried a number of config combinations with no luck

**Comment by zkochan** [comment]
Well, I am not sure why it doesn't help as I tested it locally. But seems like it is a breaking change for some: https://github.com/pnpm/pnpm/issues/4052

if this doesn't help, maybe I should revert the change

**Comment by zkochan** [comment]
Maybe the issue was in this regression: https://github.com/pnpm/pnpm/issues/4052

Try pnpm v6.23.7-202112041634

**Comment by eden-lane** [comment]
Have the same issue:
```
ERR_PNPM_FETCH_401  GET https://<...>/repository/npm-group/@babel/core/-/core-7.16.10.tgz: Unauthorized - 401
No authorization header was set for the request.
These authorization settings were found:
//nexus.redmadrobot.com/repository/npm-private:_authToken=NpmT[hidden]
always-auth=true
```

any news on this? Using pnpm 6.27.1

**Comment by ysaskia** [comment]
Hello same issues there with `pnpm 6.25.1`

```
 ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/@askia%2Fa-b: Unauthorized - 401   

An authorization header was used: Bearer ghp_[hidden]

These authorization settings were found:
@askia:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=ghp_[hidden]
```

**Comment by eden-lane** [comment]
I managed to solve the issue for Nexus Repository Manager.
My gitlab config is:
```yaml
before_script:
  - npm config set registry https:${NPM_REGISTRY}
  - echo "${NPM_REGISTRY}:_authToken=${CI_NPM_AUTH}" >> ~/.npmrc
  - echo "${NPM_REGISTRY_PRIVATE}:_authToken=${CI_NPM_AUTH}" >> ~/.npmrc
```

where `CI_NPM_AUTH` is taken from local `.npmrc` file

**Comment by brayanL** [comment]
any solution?

**Comment by fakhir-shad** [comment]
I managed to resolve it by doing a login to the registry
```
 pnpm login --registry=https://npm.pkg.github.com
```

**Comment by AustinZhu** [comment]
Same here

**Comment by kraikov** [comment]
Does anyone have a workaround? I've been trying to make this work for the last few days...

**Comment by zkochan** [comment]
What error message do you get? I have tried the configurations posted by @cybairfly and @marvin1up and pnpm picked the right tokens for me.

**Comment by zkochan** [comment]
@kraikov pnpm isn't currently reading the value from NPM_TOKEN. Here's a related feature request: https://github.com/pnpm/pnpm/issues/3141

**Comment by zkochan** [comment]
If you want to set the token via an env variable, use this env variable:

```
npm_config_//npm.pkg.github.com:_authtoken=${{ secrets.GITHUB_TOKEN }}
```

**Comment by KuhlTime** [comment]
@kraikov I got this to work inside GH Actions by adding the desired repository to the list of allowed action runners. This list is located inside your package settings page. Take this URL as reference to locate this page:

`https://github.com/USERNAME/packages/npm/PACKAGE_NAME/settings`

On this page, add the desired repository from where you want your GH Action runner to install your private package from.
 
<img width="875" alt="screenshot" src="https://user-images.githubusercontent.com/6073956/197214768-d26986d1-0996-4f1f-b59a-17b9deaf9727.png">

Once that is done, the generated `secrets.GITHUB_TOKEN` should be able to access the desired package.

For further reference, here is my `.npmrc` file as well as my workflow file I used to test this:

```
//npm.pkg.github.com/:_authToken=${NPM_TOKEN}
@OWNER:registry=https://npm.pkg.github.com
```

```yaml
name: Foo
on:
  push
env:
  NPM_TOKEN: ${{ secrets.GITHUB_TOKEN }}
jobs:
  run: 
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
         uses: actions/checkout@v3
      - name: Install Node.js
         uses: actions/setup-node@v3
         with:
           node-version: 18
      - uses: pnpm/action-setup@v2.0.1
         name: Install pnpm
         id: pnpm-install
         with:
           version: 7
```

I hope this will help you 🤞

**Comment by rajuashok** [comment]
We're doing something like this 

```
@org:registry=https://npm.pkg.github.com/org-packages
//npm.pkg.github.com/:_authToken=<token>
```

However we're getting a 404 error: 

```
 ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@flixed-io%2Fatoms: Not Found - 404

@org/package is not in the npm registry, or you have no permission to fetch it.
```

Any idea why we're seeing this?

We're also able to fetch it fine in another repo that's just using `npm` instead of `pnpm`.

**Comment by zkochan** [comment]
As you can see from the error message, pnpm requests the package from the wrong registry. Maybe it doesn't find the `.npmrc` file.

**Comment by koalazub** [comment]
Hey, I've had a look at all the issues posted with regards to trying to get this to work and I haven't been able to get it running in Actions at all. It keeps failing with the error:

>No authorization header was set for the request.
These authorization settings were found:
@owner:registry=https://npm.pkg.github.com/

My config is:
```
          echo -e "
          registry=https://registry.npmjs.org
          @owner:registry=https://npm.pkg.github.com
          //npm.pkg.github.com:_authtoken=<TOKEN>
          always-auth=true
          " > ~/.npmrc
           npm config set registry '@owner:registry' https://npm.pkg.github.com
```

At this point I've just been taking stabs in the dark and nothing really works with being able to install from the private github registry. I've also gone into each private package and [linked the repo associated with it](https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility#ensuring-workflow-access-to-your-package) as well as setting the package permissions to `read` as per the [advice up top](https://github.com/pnpm/pnpm/issues/2933#issuecomment-1287016742)

### Things I've tried
- reducing the `.npmrc` down the registries associated with the private npm packages as well
- Hard coding the tokens instead of using secrets
- setting the registry url and scope from `uses: action/setup-node`
- Updating to a later node & pnpm version(pnpm 7, node 15)

I did notice that the error output has a trailing `/`. So not sure if that's an issue again.
If anyone might spot something I may have done wrong, I'd love to have it pointed out as I'm all out of ideas with getting pnpm to work now. Otherwise, has anyone figured out a fix for themselves?

**Comment by iquirino** [comment]
Rollback pnpm version to 7.11.0 worked to me.

**Comment by zkochan** [comment]
The last two messages in this issue are probably the same as this issue: https://github.com/pnpm/pnpm/issues/5561

and you should use the solution [that I described there](https://github.com/pnpm/pnpm/issues/5561#issuecomment-1295660857).

We had to make the hard choice to deprecate always-auth in a non-major version as it was a security vulnerability.

**Comment by koalazub** [comment]
> Rollback pnpm version to 7.11.0 worked to me.

Rollback didn't work. 

> The last two messages in this issue are probably the same as this issue: https://github.com/pnpm/pnpm/issues/5561
>and you should use the solution https://github.com/pnpm/pnpm/issues/5561#issuecomment-1295660857.

The argument about the tarball location doesn't apply here as I have everything uploaded to `npm.pkg.github`. So what would be a decent way to pass auth in this case over `always-auth`? What's the ideal practice? 

I'm also seeing that my `.npmrc` has gone back to referring `registry.npmjs.org` and I'm fighting the pipeline by setting the registry. So not sure what's going on now. Is anyone else having the same issue?

**Comment by raulfdm** [comment]
I have the following `.npmrc`:

```
registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
```

And til `pnpm@7.13` it was working just fine.

I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:

```
 ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
```

Downgrading to `7.13` worked well

**Comment by koalazub** [comment]
> I have the following `.npmrc`:
> 
> ```
> registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> ```
> 
> And til `pnpm@7.13` it was working just fine.
> 
> I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:
> 
> ```
>  ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401
> 
> No authorization header was set for the request.
> 
> These authorization settings were found:
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
> ```
> 
> Downgrading to `7.13` worked well

This flip-flopping between versions has been really painful. I haven't got the bandwidth to dig in and find out why atm. Were you able to get that working in CI/CD as well? Just out of curiosity. Thanks so much for sharing

**Comment by sohrab-** [comment]
I can confirm a regression between 7.13 and 7.14, as well. If I had to take a guess, it would due to how we do auth headers now: https://github.com/pnpm/pnpm/commit/804de211ea299804b9f7d6857eeb93b618856452

**Comment by zkochan** [comment]
As far as I understand, when you use artifactory, the tarball URL differs from the metadata URL. So you need to duplicate the auth settings for the URL that hosts the tarball.

See also this issue with the same problem that was resolved by duplicating this settings as I suggest: https://github.com/pnpm/pnpm/issues/5561

In the past, it was possible to use the always-auth setting to forward auth headers to other URLs. But this was a security vulnerability, so we had to remove this setting.

**Comment by brybicki-cyferd** [comment]
See if my solution for CodeArtifact can help you guys - https://github.com/pnpm/pnpm/issues/5561#issuecomment-1298565447.

**Comment by brybicki-cyferd** [comment]
> I have the following `.npmrc`:
> 
> ```
> registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> ```
> 
> And til `pnpm@7.13` it was working just fine.
> 
> I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:
> 
> ```
>  ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401
> 
> No authorization header was set for the request.
> 
> These authorization settings were found:
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
> ```
> 
> Downgrading to `7.13` worked well

I think this will work for you:
```
registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
//company-artifactory-url.com:443/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
```

Duplicating the `_authToken` entry but including `:443` in the second one. Keep both.

**Comment by koalazub** [comment]
Alrighty, so I've managed to get everything working in GitHub Actions. Below is a list of what I've done to get it functional.

I listened to the advice of @brybicki-cyferd  https://github.com/pnpm/pnpm/issues/2933#issuecomment-1298586488 and their example with regards to implementing the registry and accounting for the tarball URL:

```
registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
//company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
//company-artifactory-url.com:443/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
```
Which afterwards gave me an error as I wasn't able to use a secret in the usual sense. So I instead opted to just implement it in an `env` like so:

```
env: 
          PAT: ${{ secrets.PAT }}
```

This was enough to get the correct authorisation to finally install my private packages. 

This was hard to navigate on a few accounts: 
- Documentation is hard to deal with currently
 - #3141 was helpful
 - #5561 helped breakdown some things related to the tarball 
- The new changes referenced regarding security had to be dug via issues and PRs. As well as that, the [release note]
- (https://github.com/pnpm/pnpm/releases/tag/v7.14.0) was too vague for me to figure out what exactly was going on. This isn't meant to come across as harsh, I only wanted to point out that the vagueness of these notes weren't able to help me solve this without pushing further by pestering people on GitHub and Discord
- The order of using an env was something I hadn't thought about. [This example](https://docs.github.com/en/actions/security-guides/encrypted-secrets#example-using-bash) helped point that out
- And finally, and this is more of an assumption, using `node 14.x` was a hindrance as well. This prompted me to make an abrupt upgrade to our node version(we're readying a big upgrade) to try and get things working. 


This was my experience with the getting things going and so I hope that this sheds some light on problems other people might be having. Thanks a bunch @zkochan for their patience and my haranguing about using pnpm.  

I'm also open to improving the way I structure my workflows. So if anyone has tips or wants to point some stuff out, then please feel free. 

<details> <summary>Workflows config:</summary>

```
name: Core Build Pipeline
on:
  pull_request:
    branches:
    - main
jobs:
  build:
    name: CI Build
    runs-on: ubuntu-latest
    timeout-minutes: 15
    steps:
      - uses: actions/checkout@v3
      - name: Use Node 15
        uses: actions/setup-node@v3
        with:
          node-version: '15.x'
          registry-url: https://npm.pkg.github.com
      - name: Set registry
        env: 
          PAT: ${{ secrets.PAT }}
        run: |
          echo "@registry_name:registry=https://npm.pkg.github.com/" >> .npmrc
          echo "//npm.pkg.github.com/:_authToken=$PAT" >> .npmrc
          echo "//npm.pkg.github.com:443/:_authToken=$PAT" >> .npmrc
      - uses: pnpm/action-setup@v2.2.4
        name: Install PNPM
        id: pnpm-install
        with:
          version: 7
          run_install: |
            - recursive: true
      - name: Build
        run: |
          pnpm run build 

```

**Comment by skimhugo** [comment]
I met  ERR_PNPM_FETCH_404  GET https://company-url/npm/npm-registry/xxxxxx: Not Found - 404

xxxxxx is not in the npm registry, or you have no permission to fetch it.

An authorization header was used: Bearer efd5[hidden]

I downgrade to 7.0 and still got error. but npm is ok.

**Comment by FlatMapIO** [comment]
The latest version of pnpm no longer works when installing `"uWebSockets.js": "github:uNetworking/uWebSockets.js#v20.15.0"`.
The terminal will show `git@github.com's password: `, and I have to move the workspace to yarn

**Comment by vasuneet** [comment]
getting same issue with 6.23.1 and 7.18.1

+ pnpm install --loglevel=verbose
 ERR_PNPM_FETCH_401  GET <Default registry>/@babel%2Fpreset-react: Unauthorized - 401

This error happened while installing a direct dependency of /opt/app/m61407_node/zlp26614.vci.att.com_27531_websocket/workspace/i-template_devdeploy_helmChanges/repository

No authorization header was set for the request.

**Comment by rikzwarthoff** [comment]
Also having this issue on 7.22.0 in combination with gitlab. 
Rolled back to 7.13.0 which is working fine.

 ERR_PNPM_FETCH_404  GET <repo-url> Not Found - 404
No authorization header was set for the request.

**Comment by slpixe** [comment]
Thanks @rikzwarthoff 
Was having same issue with `7.22.0`,
using `7.13` worked perfectly (`corepack prepare pnpm@7.13.6 --activate`)

**Comment by pix303** [comment]
Same problem with aws and CodeArtifact
`aws codeartifact login ...` ok
`pnpm publish` works fine

`aws codeartifact login ...` ok
`pnpm install @my-scope/my-package` fail with
```
:Unauthorized - 401
No authorization header was set for the request.
These authorization settings were found:
(.npmrc founded)
```

pnpm --version **7.25.0**

npm 8 works...

**Comment by goldo** [comment]
+1 
7.13.6 works and the 7.28.0 didnt work

**Comment by tlehtimaki** [comment]
> > I have the following `.npmrc`:
> > ```
> > registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> > //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And til `pnpm@7.13` it was working just fine.
> > I've upgraded to `v7.14.1` and with the same `.npmrc` I started getting:
> > ```
> >  ERR_PNPM_FETCH_401  GET https://company-artifactory-url.com:443/artifactory/api/npm/npm/@types/react-dom/-/react-dom-17.0.18.tgz: Unauthorized - 401
> > 
> > No authorization header was set for the request.
> > 
> > These authorization settings were found:
> > //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken=eyJ2[hidden]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Downgrading to `7.13` worked well
> 
> I think this will work for you:
> 
> ```
> registry = https://company-artifactory-url.com/artifactory/api/npm/npm/
> //company-artifactory-url.com/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> //company-artifactory-url.com:443/artifactory/api/npm/npm/:_authToken = ${ARTIFACTORY_TOKEN}
> ```
> 
> Duplicating the `_authToken` entry but including `:443` in the second one. Keep both.

I found that duplicating `_authToken_` line with port `:443` worked for me, but I think its perhaps a separate issue.

I got into to the situation like this:

1. Init new project with pnpm 
2. Install some packages from public npm regisry
3. Notice you now need stuff from private npm registry
4. Setup everything for that
5. Now try to install private package
6. It wont work. The tarball has `:443` in the URL so I suspect that could have something to do with the error
7. Add the extra `_authToken` line to your `.npmrc` config.
8. Now install the private package.
9. Package is installed succesfully

**Comment by goldo** [comment]
thanks @raulfdm it fixed it

**Comment by viresh323** [comment]
> Just complementing what @tlehtimaki said, on my team, we had problems with per project `.npmrc`. They still get those errors even if they add the `_authToken`.
> 
> The solution was by also adding the `_authToken` for port 443 in the root level `.npmrc` (for UNIX users it may be located at `~/.npmrc`. Or check it here: https://pnpm.io/cli/config)

Thanks, I had same problem with the port.

**Comment by joostdelange** [comment]
Good news for CodeArtifact users, from AWS' side this problem has been solved. Quoting a CodeArtifact operation notification all customers who use it got:

> You are receiving this notification because your account has recently downloaded an npm or NuGet package from an AWS CodeArtifact repository.

> For npm and NuGet packages, CodeArtifact returns URLs in the HTTP response body for certain request types. For example, when the `npm install` command is run, the npm client will request metadata for each package it needs to install. This metadata contains CodeArtifact URLs that the client will request in order to complete the installation. The same pattern is also used by NuGet clients such as the NuGet CLI and dotnet CLI, and alternate npm clients such as Yarn and pnpm.

> Currently these returned CodeArtifact URLs include the HTTPS port number (443) after the hostname, for example [1].

> Starting March 27, 2023, CodeArtifact will no longer include the port number in returned URLs, for example [2].

> This change will improve compatibility with the pnpm client and make the returned URLs consistent with those returned by the CodeArtifact `GetRepositoryEndpoint` API. No backward compatibility issues are expected for npm or NuGet clients as a result of the change as the URLs with and without the port number are functionally identical.

> There is no impact to Maven and Python packages as these repository protocols do not include CodeArtifact URLs in any response.

> If you have any questions or concerns, please reach out to AWS Support [3].

> [1] [https://domain-123456789012.d.codeartifact.us-west-2.amazonaws.com:443/npm/repo/lodash/-/lodash-4.17.21.tgz](https://domain-123456789012.d.codeartifact.us-west-2.amazonaws.com/npm/repo/lodash/-/lodash-4.17.21.tgz)
[2] https://domain-123456789012.d.codeartifact.us-west-2.amazonaws.com/npm/repo/lodash/-/lodash-4.17.21.tgz
[3] https://aws.amazon.com/support

**Comment by adcorduneanu** [comment]
I want to add one extra scenario I encountered that does not make any sense to me - the URL is case-sensitive...

Config:
```ini
registry=http://xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/
always-auth=true
shamefully-hoist=true
strict-ssl=false

; begin auth token
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/:username=${WWW_NPM_USER}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/:_password=${WWW_NPM_TOKEN}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/registry/:email=npm requires email to be set but doesn't use the value
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/:username=${WWW_NPM_USER}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/:_password=${WWW_NPM_TOKEN}
//xxx:8080/tfs/www_bu_mmm/_packaging/npmProductionFeed/npm/:email=npm requires email to be set but doesn't use the value
; end auth token
```

Issue:
```ini
Packages: +851
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 ERR_PNPM_FETCH_401  GET http://xxx:8080/tfs/Www_BU_Mmm/_packaging/npmProductionFeed/npm/registry/@babel/core/-/core-7.20.12.tgz: Unauthorized - 401

No authorization header was set for the request.

These authorization settings were found:
```

**Comment by maximLyakhov** [comment]
Got this error within renovate bot job.
Solved by using _npmrc_ property in **config.js** with the help of this thread:

```
module.exports = {
  npmrc: `
    @${process.env.COMPANY}:registry=https://gitlab.${process.env.DOMAIN}.com/api/v4/projects/{process.env.PROJECT_ID}/packages/npm/
    //gitlab.${process.env.DOMAIN}.com/api/v4/projects/${process.env.PROJECT_ID}/packages/npm/:_authToken = ${process.env.CI_NPM_TOKEN}
    //gitlab.${process.env.DOMAIN}.com:443/api/v4/projects/${process.env.PROJECT_ID}/packages/npm/:_authToken = ${process.env.CI_NPM_TOKEN}
  `
},
```
Thanks for everyone's input.

**Comment by nielsbrakel** [comment]
For everyone running into this issue in **Azure Devops** we solved it by including the following command in our pipeline, hopefully this is useful for someone 😊

```yaml
  - task: npmAuthenticate@0
    inputs:
      workingFile: $(WorkingDirectory)/.npmrc
```

Source:
https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/npm-authenticate-v0?view=azure-pipelines

**Comment by dtrongphuc** [comment]
In the case Gitlab Group Registry. This work well for me:

```shell
@scope:registry=https://gitlab.com/api/v4/groups/${GROUP_ID}/-/packages/npm/
//gitlab.com/api/v4/projects/${PROJECT_ID1}/:_authToken=${GROUP_TOKEN}
//gitlab.com/api/v4/projects/${PROJECT_ID2}/:_authToken=${GROUP_TOKEN}
```

**Comment by manawasp** [comment]
We got private package resolving issues too and the last valid version is `7.13.6`.

Some context: we use a Gitlab instance on premise and our .npmrc look like:

```
//gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN
@team-software:registry=https://gitlab.domain.com/api/v4/packages/npm/
```

We try to install a library hosted on this gitlab registry: `"@team-software/v-components": "0.0.96"`

This configuration works with `pnpm<=7.13.6` or `npm`  but with pnpm `8.6.6` (and node `20.3.0`) we get:

```
 WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 10 seconds. 2 retries left.
 WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 1 minute. 1 retries left.
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /builds/.pnpm-store/v3
  Virtual store is at:             node_modules/.pnpm
 ERR_PNPM_FETCH_404  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz: Not Found - 404
 No authorization header was set for the request.
```

I did some tests:

- I added `gitlab.domain.com/api/v4/projects/:_authToken=AUTH_TOKEN` and this fixed the issue
- Then I removed `//gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN` this brings a new resolving issue. I guess no match between `@team-software:registry=xxx` and the `xxx:_authToken=token`
- I merged both line and used  `//gitlab.domain.com/api/v4/:_authToken=AUTH_TOKEN`,  **this is working** (but doesn't follow what gitlab recommend to use for npm).

Final:

I feel like to resolve the package there is a first match on `gitlab.domain.com/api/v4/packages/npm/` then a download issued on the url result  `gitlab.domain.com/api/v4/projects/`. Before the header `AUTH_TOKEN` issued on the first match was forwarded on the "related download" even if there was a matching mismatch?

_Just a guess without diving too much how things are working, if someone has more context about this 🙏_

**Comment by slpixe** [comment]
@manawasp massive thanks for this, this has solved my issue of getting recent versions of pnpm to work with a private gitlab.com repository

The main line which got it working was appending
`//gitlab.com/api/v4/:_authToken=AUTH_TOKEN`

**Comment by MonstraG** [comment]
So, how does one install (not even trying publishing) in github actions now?

I'm on pnpm@8.7.0, and trying to use https://gist.github.com/belgattitude/838b2eba30c324f1f0033a797bab2e31, with added `env`:
```yml
    - name: Install dependencies
      shell: bash
      working-directory: ${{ inputs.cwd }}
      run: pnpm install --frozen-lockfile --prefer-offline
      env:
        NPM_TOKEN: ${{ github.token }}
        NPM_AUTH_TOKEN: ${{ github.token }}
        NODE_AUTH_TOKEN: ${{ github.token }}
        npm_config__authtoken: ${{ github.token }}
        npm_config_//npm.pkg.github.com:_authtoken: ${{ github.token }}
```
.npmrc:
```
resolution-mode=highest
strict-peer-dependencies=true
auto-install-peers=false
@my-cool-repo:registry=https://npm.pkg.github.com
```

I get:
```
ERR_PNPM_FETCH_401  GET https://npm.pkg.github.com/download/...: Unauthorized - 401
No authorization header was set for the request.
These authorization settings were found:
@lifekeys:registry=https://npm.pkg.github.com/
_authToken=ghs_[hidden]
```

I cannot add `//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}` because then I get `WARN Issue while reading  "/home/..../.npmrc". Failed to replace env in config: ${GITHUB_TOKEN}` when installing locally, and that makes pnpm ignore `auto-install-peers` setting there.

I cannot downgrade pnpm to `7.13.6` in the `action.yml` because option to use lockfile v6 was added in `7.24.0`

what do I do?

**Comment by SpeedySH** [comment]
> We got private package resolving issues too and the last valid version is `7.13.6`.
> 
> Some context: we use a Gitlab instance on premise and our .npmrc look like:
> 
> ```
> //gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN
> @team-software:registry=https://gitlab.domain.com/api/v4/packages/npm/
> ```
> 
> We try to install a library hosted on this gitlab registry: `"@team-software/v-components": "0.0.96"`
> 
> This configuration works with `pnpm<=7.13.6` or `npm` but with pnpm `8.6.6` (and node `20.3.0`) we get:
> 
> ```
>  WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 10 seconds. 2 retries left.
>  WARN  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz error (ERR_PNPM_FETCH_404). Will retry in 1 minute. 1 retries left.
> Packages are hard linked from the content-addressable store to the virtual store.
>   Content-addressable store is at: /builds/.pnpm-store/v3
>   Virtual store is at:             node_modules/.pnpm
>  ERR_PNPM_FETCH_404  GET https://gitlab.domain.com/api/v4/projects/17/packages/npm/@team-software/v-components/-/@team-software/v-components-0.0.96.tgz: Not Found - 404
>  No authorization header was set for the request.
> ```
> 
> I did some tests:
> 
> * I added `//gitlab.domain.com/api/v4/projects/:_authToken=AUTH_TOKEN` and this fixed the issue
> * Then I removed `//gitlab.domain.com/api/v4/packages/npm/:_authToken=AUTH_TOKEN` this brings a new resolving issue. I guess no match between `@team-software:registry=xxx` and the `xxx:_authToken=token`
> * I merged both line and used  `//gitlab.domain.com/api/v4/:_authToken=AUTH_TOKEN`,  **this is working** (but doesn't follow what gitlab recommend to use for npm).
> 
> Final:
> 
> My understanding is to resolve the package there is a first match on `gitlab.domain.com/api/v4/packages/npm/` then a download issued on the url result `gitlab.domain.com/api/v4/projects/`. Before the header `AUTH_TOKEN` issued on the first match was forwarded on the "related download" even if there was an uri mismatch?
> 
> _Just a guess without diving too much how things are working, if someone has more context about this 🙏_

You saved me. I spent 3 hours to find this answer and it really works.

**Comment by ekashida** [comment]
It sounds like this is a known issue and is described at length here:
https://github.com/npm/cli/wiki/%22No-auth-for-URI,-but-auth-present-for-scoped-registry%22

**Comment by weyert** [comment]
Try adding the one with the project id in your config, e.g.
```
gitlab.domain.com/api/v4/projects/17/packages/npm/:_authToken=AUTH_TOKEN
```
For me adding all these urls for each package solved all these issues. Sure, I have 20+ urls in my `.npmrc` file but no problems :)

I think i saw that my ticket at Gitlab recently got updated/resolved so might want to try in last months/this month Gitlab release or on gitlab.com.

**Comment by tlehtimaki** [comment]
> It sounds like this is a known issue and is described at length here: https://github.com/npm/cli/wiki/%22No-auth-for-URI,-but-auth-present-for-scoped-registry%22

Would be interesting to hear if widely if the npm related issue has helped people with this too.

**Comment by novo-esdb** [comment]
Have anyone had similar issues trying to access a Azure DevOps artifact feed? 

I am getting issues similar to this:

```
 ERR_PNPM_FETCH_401  GET https://pkgs.dev.azure.com/<org>/_packaging/<project>/npm/registry/@<org-alias><project>: Unauthorized - 401

An authorization header was used: Bearer ejVh[hidden]

These authorization settings were found:
//pkgs.dev.azure.com/<org>/_packaging/<project>/npm/:_authToken=ejVh[hidden]
//pkgs.dev.azure.com/<org>/_packaging/<project>/npm/registry/:_authToken=ejVh[hidden]
@<org-alias>:registry=https://pkgs.dev.azure.com/<org>/_packaging/<project>/npm/registry/
```

where my local `.npmrc` has the authentication settings shown above.

**Comment by rishavanand** [comment]
thanks @danielbayley with pnpm 

adding a slash worked

before
```
//npm.pkg.github.com:_authToken=***
```

after
```
//npm.pkg.github.com/:_authToken=***
```

**Comment by zkochan** [comment]
A PR has been submitted that should hopefully fix this: https://github.com/pnpm/pnpm/pull/7337

**Comment by bilalafzal01** [comment]
I just had to configure pnpm as well (with `pnpm config set`) rather than just having `.npmrc`


github workflow:
```
name: Publish artifacts
jobs:
  deploy:
    name: Deploy CDK
    runs-on: ubuntu-latest
    env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    permissions:
        packages: read
        contents: read
    steps:
        - name: Configure PNPM
          run: pnpm config set '//npm.pkg.github.com/:_authToken' "${GITHUB_TOKEN}"
  
        - name: Install dependencies
          run: cd frontend && pnpm install
```

.npmrc:
```
@NAMESPACE:registry=https://npm.pkg.github.com
```

**Comment by revskill10** [comment]
Same issue.

**Comment by aiell0** [comment]
I was using this in a docker container, and was getting 401 errors even with the token set correctly. I had to add the following to my `~/.npmrc` file:

`RUN echo "registry=https://npm.pkg.github.com/<owner>" >> ~/.npmrc`

**Comment by yamcodes** [comment]
Hey guys! None of the above solutions we tried worked for us for **PRivate GitHub Packages** and **GitHub Actions**.
After much tinkering, here's what worked (tested in pnpm@v8):

```yml
      - uses: pnpm/action-setup@v3
        name: Install pnpm
        with:
          version: 8

      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: "lts/*"
          cache: "pnpm"
          scope: "@ourscope"
          registry-url: "https://npm.pkg.github.com"

      - name: Configure PNPM auth for private registry
        run: pnpm config set //npm.pkg.github.com/:_authToken=${{ secrets.SOME_PAT }}

      - name: Install dependencies
        run: pnpm install
```

Whereas `SOME_PAT` is a secret for a Personal Access Token with `read:packages`, and `@ourscope` is our scope name for our private GitHub Packages registry.

We don't have anything related to auth or registries in `.npmrc`, so that part wasn't necessary for us.
